util/topology/topology.go

```diff
@@ -0,0 +1,51 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package topology implements topology utility functions.
+package topology
+
+import (
+	"context"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+// ShouldSkipImmutabilityChecks returns true if it is a dry-run request and the object has the
+// TopologyDryRunAnnotation annotation set, false otherwise.
+func ShouldSkipImmutabilityChecks(ctx context.Context, obj metav1.Object) bool {
+	req, err := admission.RequestFromContext(ctx)
+	if err != nil {
+		return false
+	}
```

Comment:
I'm not sure if this should be inside of this func
1. Calling it inside of a CustomValidator instead guarantees that the error does not occur (given a minimum CR version)
2. This locks the util func itself to a specific (minimum) controller-runtime version. While keeping it outside of it only couples it to the admission API
3. The function signature looks too magical to me. Feels way more explicit to pass in the admission.Request
    * This would also allow folks to use this func in a relatively normal way by implementing the webhook directly with the lower-level CR API (admission.Handler)

Of course the downside would be that folks have to add a few more lines to their webhook


cc @vincepri WDYT?

(but no showstopper from my side)

Reply:
I'd be okay with both :-). Will move it up for now.